PRINCIPLES.md

```diff
@@ -0,0 +1,254 @@
+
+All versions, except to very first,  should also have reference a predecessor or parent, which is another uniquely named version. This enables us to retain a history of the changes.
+
+<!--
+#### Why is it necessary to have versions be immutable and retained indefinitely?
+
+
+
+- only verb is to create a new version.
+This - Only uniquely named new versions of the system configuration can be created and all configuration info is retained (forward only)
+- rollback
+- Audit
+The only exception being in the case of sensitive information leaks due to error.
+  (I would agrue that key rotation is always better option in this case, but for PII and proprietary info this isn't possible).
+- changes as transactions
+
+#### What are the responsibilities of a State store?
+- In addition, state stores must preserve metadata information regarding the identity of agents that create new versions.
+The Desired State of the new version, Who created it, when it was created and ideally, also _why_ it was created. - What Why, who when. 
+
+- Even though we say git, it's actually only true when configured in a very particular way.
+
+#### State stores as coordination points/ collab via common data interface
+- Collab point. 
+- Business rules and security rules about the acceptable state of the system shoudl be checked against any new proposed version of the desired configuration before such changes are accepted as a new version. (Gates before version)
+- where human process is integrated
+- the place where collaboration can be synchronised.
+-->
+
+### 3. Continuous state reconciliation
+
+<div style="background: #E3F2FD; border-left: 3px solid #0D47A1; padding:10px; margin-bottom:5px; font-style:italic">
+Software agents continuously, and automatically, compare a system's <em>Actual State</em> to its <em>Desired State</em>.
+If the actual and desired states differ, automated actions are immediately attempted to reconcile them.
+These differences could be due to the actual state drifting from the desired state, or the desired state changing intentionally.
+</div>
+
+<!--
+- If the software agents fail to bring the system's state in line with its desired state, a human operator is notified.
+- Software agents continuously check that the running system under management matches the desired state configuration, and if it does not, immediately either take remedial action to bring the system back in line with stated expectations or, if this cannot be done, alert a human operator that the system is no longer meeting expectations
+- Automated delivery: Delivery of the declarative descriptions, from the repository to runtime environment, is fully automated.
+- Software Agents: Reconcilers maintain system state and apply the resources described in the declarative configuration.
+- Closed loop: Actions are performed on divergence between the version controlled declarative configuration and the actual state of the target system.
```

Comment:
When I read closed loop I immediately think of https://en.wikipedia.org/wiki/Control_theory#Open-loop_and_closed-loop_(feedback)_control 

Most GitOps implementations are open reconciliation loops, the change is applied irrespective of the target state and the success or failure of the operation does not feedback into Git.  Some GitOps products e.g. Argo do try and build closed loops based but their feedback is internal only (either via UI or Kuebrentes events)  - Will open a new PR on this topic as I think it is an unsolved problem

Reply:
It's true the _Current State_ doesn't feedback to the _Desired State_ store, but it does feed back to the _Software Agents_ so they know how the current and desired states do or do not diverge. This part of that wikipedia page you linked sounds like a correct description to me, where the "controller" map to GitOps _Software Agents_ :

> The definition of a closed loop control system according to the British Standard Institution is "a control system possessing monitoring feedback, the deviation signal formed as a result of this feedback being used to control the action of a final control element in such a way as to tend to reduce the deviation to zero."

WDYT?